File: recursion/max_number.rb
=begin

[1, 2, 4, 7, 3]
- compare 1 and 2
- 2 is max
- compare 2 and 4
- 4 is max
- compare 4 and 7
- 7 is max
- compare 7 and 3
- 7 is max

- base case is pointer is after last index

=end

def max_number(array, pointer=0, max=0)
  return max if pointer == array.size

  max = array[pointer] if array[pointer] > max
  max_number(array, pointer + 1, max)
end

# ============================================
#
# without pointer approach:

def max_number(array, max=0)
  return max if array.size == 0

  max = array.first if array.first > max
  max_number(array[1..-1], max)
end

# ===========================================
# 
# answer from group:

def max_number(nums, max=-Float::INFINITY)
  if nums.length == 1
    max = nums.first if nums.first > max
    return max
  else
    mid = (nums.length - 1) / 2
    left = nums[0..mid]
    right = nums[mid + 1..nums.length - 1]
  end

  max = max_number(left, max)
  max = max_number(right, max)
  max
end

puts max_number([1, 2, 4, 7, 3]) == 7
puts max_number([1, 8, 4, 7, 3, -2]) == 8
puts max_number([1]) == 1

File: recursion/max_number.js
// Find the maximum number in an array of numbers. Return that number.
//
// left/right recursion

var max_number = function(nums) {
  if (nums.length === 1) { return nums[0]; }
  if (nums.length === 2) {
    return nums[0] > nums[1] ? nums[0] : nums[1];
  }

  var mid = Math.floor((nums.length - 1) / 2);
  var left_max = max_number(nums.slice(0, mid + 1));
  var right_max = max_number(nums.slice(mid + 1, nums.length));

  return left_max > right_max ? left_max : right_max;
};

console.log(max_number([1, 3, 2, 4]) === 4);
console.log(max_number([1, 3, 2, 4, 8, 10, -2]) === 10);
console.log(max_number([11, 1, 3, 2, 4, 8, 10, -2]) === 11);
console.log(max_number([11, 1, 3, 2, 4, 8, 11, 10, -2]) === 11);
console.log(max_number([1]) === 1);
console.log(max_number([1, -5]) === 1);
